plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,482 @@
+						      &ctx->error)) {
+		g_debug ("error: couldn't parse response message: %s\n", ctx->error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	g_debug ("[%s] Successfully request modem to query caps", mbim_device_get_path_display (device));
+
+	ctx->firmware_version = g_strdup (firmware_version);
+
+	g_main_loop_quit (ctx->mainloop);
+}
+
+static void
+fu_mbim_qdu_updater_caps_query (MbimDevice *device, GetFirmwareVersionContext *ctx)
+{
+	g_autoptr(MbimMessage) request = NULL;
+
+	request = mbim_message_device_caps_query_new (NULL);
+
+	mbim_device_command (device,
+			     request,
+			     10,
+			     NULL,
+			     (GAsyncReadyCallback) fu_mbim_qdu_updater_caps_query_ready,
+			     ctx);
+}
+
+gchar *
+fu_mbim_qdu_updater_check_ready (FuMbimQduUpdater *self, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	GetFirmwareVersionContext ctx = {
+		.mainloop = mainloop,
+		.error = NULL,
+		.firmware_version = NULL,
+	};
+
+	fu_mbim_qdu_updater_caps_query (self->mbim_device, &ctx);
+
+	g_main_loop_run (mainloop);
+
+	if (ctx.error != NULL) {
+		g_propagate_error (error, ctx.error);
+		return NULL;
+	}
+
+	return ctx.firmware_version;
+}
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+	GError		*error;
+	GBytes		*blob;
+	GArray		*digest;
+	GPtrArray	*chunks;
+	guint 		 chunk_sent;
+	FuDevice 	*device;
+} WriteContext;
+
+static gboolean
+fu_mbim_qdu_updater_reboot_timeout (gpointer user_data)
+{
+	WriteContext *ctx = user_data;
+
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_mbim_qdu_updater_file_write_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	WriteContext *ctx = user_data;
+	g_autoptr(MbimMessage) response = NULL;
+
+	response = mbim_device_command_finish (device, res, &ctx->error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &ctx->error)) {
+		g_debug ("error: operation failed: %s\n", ctx->error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_write_response_parse (response, &ctx->error)) {
+		g_debug ("error: couldn't parse response message: %s\n", ctx->error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	fu_device_set_progress_full (FU_DEVICE (ctx->device), (gsize) ctx->chunk_sent, (gsize) ctx->chunks->len);
+	ctx->chunk_sent++;
+	if (ctx->chunk_sent < ctx->chunks->len) {
+		FuChunk *chk = g_ptr_array_index (ctx->chunks, ctx->chunk_sent);
+		g_autoptr(MbimMessage) request = mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk),
+										      (const guint8 *) fu_chunk_get_data (chk), NULL);
+		mbim_device_command (ctx->mbim_device,
+				     request,
+				     10,
+				     NULL,
+				     (GAsyncReadyCallback) fu_mbim_qdu_updater_file_write_ready,
+				     ctx);
+		return;
+	}
+
+	fu_device_set_progress (FU_DEVICE (ctx->device), 100);
+	g_debug ("Done!");
```

Comment:
nuke this i think

Reply:
Sorry, do you mean to modify it more detail or just remove it?

Reply:
remove please